src/yki/handler/paytrail_payment_report.clj

```diff
@@ -0,0 +1,83 @@
+(ns yki.handler.paytrail-payment-report
+  (:require
+    [clj-time.core :as t]
+    [clojure.data.csv :as csv]
+    [clojure.data.json :as json]
+    [clojure.java.io :as jio]
+    [clojure.string :as str]
+    [clojure.tools.logging :as log]
+    [compojure.api.sweet :refer [api context POST]]
+    [integrant.core :as ig]
+    [ring.middleware.params :refer [wrap-params]]
+    [ring.util.http-response :refer [ok]]
+    [yki.handler.routing :as routing]
+    [yki.middleware.payment :refer [with-request-validation]])
+  (:import (java.io FileOutputStream InputStream)))
+
+(defn- infer-content-type [headers]
+  (let [content-type-string (headers "content-type")]
+    (when content-type-string
+      (cond
+        (str/starts-with? content-type-string "text/csv")
+        :csv
+        (str/starts-with? content-type-string "application/json")
+        :json))))
+
+(defn report-file-path ^String [content-type]
+  (case content-type
+    :csv
+    (str "/tmp/yki/report/csv/" (t/now) ".csv")
+    :json
+    (str "/tmp/yki/report/json/" (t/now) ".json")))
+
+(defn- store-report! [content-type report-contents]
+  (let [path-to-report (report-file-path content-type)]
+    (jio/make-parents path-to-report)
+    (with-open [fos (FileOutputStream. path-to-report)]
+      (if (instance? InputStream report-contents)
+        (.transferTo ^InputStream report-contents fos)
+        ; Else report-contents is likely a map or string
+        ; -> try to just spit it.
+        (spit fos report-contents)))))
+
+(defn content->records [content-type report-contents]
+  (log/info "Trying to interpret records of type" content-type "with report-contents of class" (class report-contents))
+  (case content-type
+    :csv
+    (let [[header & rows] (csv/read-csv report-contents :separator \;)]
+      (log/info "CSV payload contains the following fields:" (str/join " | " header))
+      (map zipmap (repeat header) rows))
+    :json
+    (json/read-str report-contents)))
+
+(defmethod ig/init-key :yki.handler/paytrail-payment-report [_ {:keys [auth payment-helper]}]
+  {:pre [(some? auth) (some? payment-helper)]}
+  (api
+    (context routing/paytrail-payment-report-root []
+      :coercion :spec
+      :no-doc true
+      :middleware [wrap-params #(with-request-validation (:payment-config payment-helper) %)]
```

Comment:
`auth` ei parhaillaan käytössä, lisäiskö middlewaren alle ja jos/kun nää rajapinnat tarkoitus olla julkisia(?) niin sinne `any-access` näille `/payment-report/paytrail` alaisille rajapinnoille?

Reply:
Sinänsä toi `auth` nyt on tässä hieman turha kun autentikointi tapahtuu oikeasti tuon `with-request-validation`-funktion avulla. Authin voisi siis ehkä ottaa kokonaan pois täältä.

Reply:
Selkeyden kannalta olisi ehkä parasta siirtää with-request-validation osaksi noita `yki.middleware.auth` alla listattuja access ruleja, jolloin Paytrail-kutsujen autentikointi voisi hieman yksinkertaistua handlereissa (tämä, exam-payment-new ja evaluation-payment-new). Vaatisi vain pientä jumppausta ja trial-and-erroria että saa varmasti toimimaan sen oikein; muistaakseni esim. noiden middlewarejen järjestyksen kanssa voi tässä joutua hieman venkslaamaan.

Reply:
Okei! En joo alkanut tutkimaan mitä tuon `with-request-validation` alta löytyy. authin poisto tässä kohtaa myös ok.